### PR TITLE
Add a new query type - ScriptScoreQuery

### LINKED ISSUE #27588: Replacing the function_score with discrete queries
## Overview
The `function_score` is a powerful query, but can be somewhat unwieldy and difficult to use.  It is a monolithic query that has many parameters and options which makes it difficult for new users to learn.  It can be difficult to tweak, and now that we've moved to BM25, the defaults don't work well (namely, multiplying scores).

We'd like to deprecate `function_score` and replace it's functionality with a number of discrete queries.  The guiding principle is that these replacement queries will be small, simple and single-purpose.  Individual queries should be simpler to implement and maintain, easier for users to learn and should allow the same functionality as `function_score` by mixing/matching as required.  

#### Arbitrary numeric functions 
This query will allow working with arbitrary numerics and allow applying various mathematical functions.  For example, a document may have a `popularity` field that you wish to roll into the score somehow.

Functions should include those of `field_value_factor`, like logarithm, sqrt, reciprocal, etc.  We'd also like to include a sigmoid and rational function.  The query should also be able to return just the value itself.  I think the random number functionality of `function_score` can be rolled into this too.

We could potentially implement this query entirely through scripting, assuming we can provide custom functions through the script context for more complicated operations like sigmoid.

#### Distance: Numeric and Time
This query would provide a set of decay functions that allow you to score the "distance" from a field value to some point.  When dealing with dates this "distance" would essentially represent recency, while with numerics it'd just be a geometric distance.

We'll likely want numeric/time to be grouped together since the operations are essentially the same.

#### Distance: Geo
This would be similar to numeric/time distance, except operating on geo points and physical distance.  The thinking is that geo would be separate, as we may have plans in the future for geo querying to become more robust in general.  Geo points are also sufficiently different that the syntax will likely need to be a bit different. 

Otherwise the functionality is similar, providing a set of decay functions.

## Potential queries

Some other potential queries that we kicked around, unsure if they are needed or quite how the functionality would work.  Some of these have a direct comparison in function_score, some are only tangentially related.

#### Min/Max query
Function score allows taking the min or max score from the set of functions.  We could potentially add a combination query that executes the children, then only passes on the min/max score from the children.

#### Bandpass query / cutoff score query
Function score allows setting max thresholds on scores generated by the functions.  It may be useful to have a query that allows setting min or max or both, and the scores that come out of the query would be limited to those values.  Similar to a constant score in that it wraps a set of queries, but instead of setting a single score it just limits the scores that are generated to the range.

Note this is different from the above min/max query, in that it limits the produced score to a range, whereas min/max simply takes the min or max score as-is.

Would allow us to remove `min_score` (https://github.com/elastic/elasticsearch/issues/13115), and limit individual queries (https://github.com/elastic/elasticsearch/issues/17348)

#### In-order Boolean
One unique aspect of `function_score` is the ability to use only the "first" matching function.  There's no other place in the query DSL that allows "short-circuiting" evaluation... the equivalent "first" functionality would require a complex set of must/must_not boolean conditions.

So it may be useful to implement an "in-order" boolean query which evaluates child queries in their order and allows exiting after some criteria is met (`first`, etc).

It's not entirely clear how useful this functionality is outside of the `function_score` though, and we'd be interested to hear use-cases for this `first` behavior.

#### Scripted Boolean
Related to all the above, it may be useful to have a boolean that executes all the child queries and provides those scores to a script, which would then decide how to combine the scores.  This could allow very sophisticated behavior by allowing the user to script away which scores are included (e.g. if they meet a criteria, or only if the total boolean exceeds a threshold, etc) and how they are combined

The downside is that all child queries must be run so that all scores can be collected.  And the syntax/script interface would likely be complex

## Related issues
https://github.com/elastic/elasticsearch/issues/23850
https://github.com/elastic/elasticsearch/issues/15670 (by boosting each individual decay query)

/cc @mayya-sharipova @colings86 @clintongormley did I miss anything?

### PR DESCRIPTION
**Pull request by mayya-sharipova**
script_score query uses script to calculate document scores.
Added as a substitute for function_score with an intention
to deprecate `function_score` query.

```http
GET /_search
{
    "query": {
        "script_score" : {
            "query": {
                "match": { "message": "elasticsearch" }
            },
            "script" : {
              "source": "_score + Math.log(2 + doc['likes'].value)"
            },
            "min_score" : 2
        }
    }
}
```

Add several functions to painless to be used inside script_score:

double rational(double, double)
double sigmoid(double, double, double)

double decayGeoLinear(String, String, String, double, GeoPoint)
double decayGeoExp(String, String, String, double, GeoPoint)
double decayGeoGauss(String, String, String, double, GeoPoint)

double decayNumericLinear(double, double, double, double, double)
double decayNumericExp(double, double, double, double, double)
double decayNumericGauss(double, double, double, double, double)

double decayDateLinear(String, String, String, double, JodaCompatibleZonedDateTime)
double decayDateExp(String, String, String, double, JodaCompatibleZonedDateTime)
double decayDateGauss(String, String, String, double, JodaCompatibleZonedDateTime)

Date functions only works on dates in  the default format and default time zone

Relates to #23850
Closes #27588, #30303

### CODE REVIEW

**Comment by Conky5** [comment]
jenkins test this please

File: server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+     */
+    public ScriptScoreFunctionBuilder scriptScoreFunctionBuilder() {
+        return this.scriptScoreFunctionBuilder;
+    }
+
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (query != null) {
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java]: Is this check necessary? We have a null check for query in the constructor so I don't think this can be null here?

File: server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    protected int doHashCode() {
+        return Objects.hash(this.query, this.scriptScoreFunctionBuilder, this.minScore);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        ScriptScoreFunction function = (ScriptScoreFunction) scriptScoreFunctionBuilder.toFunction(context);
+        Query query = this.query.toQuery(context);
+        if (query == null) {
+            query = new MatchAllDocsQuery();
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java]: What conditions lead to the `toQuery()` call emitting `null`? I am concerned that assuming `null` can be converted to a `MatchAllDocsQuery` might be trappy here

File: server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        QueryBuilder query = null;
+        ScriptScoreFunctionBuilder functionBuilder = null;
+        String queryName = null;
+        Float minScore = null;
+        float boost = AbstractQueryBuilder.DEFAULT_BOOST;
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java]: Is there a reason this parsing can't be done using either `ObjectParser` or `ConstructingObjectParser`? I think if we can do it with one of those it would be great as its easier to see what they are doing

**Review by colings86 — COMMENTED** [review]
@mayya-sharipova I left some comments.

**Comment by mayya-sharipova** [comment]
@colings86 Colin, thanks so much for the review, I will address your comments shortly

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java [hunk 4]
```diff
@@ -154,6 +155,33 @@ public boolean needs_score() {
                 }
             };
             return context.factoryClazz.cast(factory);
+        } else if (context.instanceClazz.equals(ScoreScript.class)) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java]: I'm confused on why this case is added. I see a couple issues with it:
* GenericElasticsearchScript is about to be removed (see https://github.com/elastic/elasticsearch/pull/34730), and all the uses of it (through SearchScript) have already been removed
* We already have caching of the compiled script in ScriptService, so why do we need special handling of caching here?

**Comment by mayya-sharipova** [comment]
@colings86 @rjernst I have pushed new commits that address your feedback, can you please continue to review when you have available time

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.lucene.search.Scorer;
+import org.elasticsearch.ElasticsearchException;
+
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A query that allows for
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: incomplete javadoc?

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java [hunk 6]
```diff
@@ -77,6 +83,11 @@ public void setScorer(Scorable scorer) {
         };
     }
 
+    /** Return the score of the current document. */
+    public double getScore() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java]: Why is this added? This would make `score` available in addition to `_score` (which is already available with `get_score()` below). Why do we need the additional score accessor, that deviates from the access in other scripts?

**Review by rjernst — COMMENTED** [review]
This script changes look good to me now. Just a couple small nits.

**Comment by mayya-sharipova** [comment]
@colings86 @rjernst  I have tried to address your comments. Can you please have another round of review, when you have available time?

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A query that uses a script to compute documents' scores.
+ */
+public class ScriptScoreQuery extends Query {
+    final Query subQuery;
+    final ScriptScoreFunction function;
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: Given that the ultimate aim is to remove the Function Score Query entirely it would be great if we could make it so this query doesn't rely on the ScriptScoreFunction either and inlines that logic. I don't think thats completely necessary for this PR but we should consider doing it in a follow up

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        this.subQuery = subQuery;
+        this.function = function;
+        this.minScore = minScore;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query rewritten = super.rewrite(reader);
+        if (rewritten != this) {
+            return rewritten;
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java]: In most of the Lucene query implementations the call to `super.rewrite()` is done on the final return line (in this case line 66) so I wonder if its worth doing the same here for readability and consistency? I think the logic would end up being the same we would just be checking the internals of this  subclass first?

File: server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        }
+        return this;
+    }
+
+    @Override
+    protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
+        InnerHitContextBuilder.extractInnerHits(query(), innerHits);
+    }
+
+    private static ConstructingObjectParser<ScriptScoreQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME, false,
```
Comment (colings86) [inline server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java]: nit: Could we put the static blocks, methods and constants at the top of the class?

**Review by colings86 — COMMENTED** [review]
@mayya-sharipova I left a few more comments but I think this is close assuming @rjernst is happy with the changes on the scripting side

**Comment by mayya-sharipova** [comment]
@rjernst @colings86  I have tried to address your feedback. Would you have any other comments for this PR?

File: docs/reference/query-dsl/script-score-query.asciidoc [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+}
+--------------------------------------------------
+// NOTCONSOLE
+
+or set your own reproducible seed using this script:
+
+[source,js]
+--------------------------------------------------
+"script" : {
+    "source" : "Random rnd = new Random(); rnd.setSeed(doc['field'].value); rnd.nextFloat()"
```
Comment (colings86) [inline docs/reference/query-dsl/script-score-query.asciidoc]: I think this should use a seed from a parameter rather than from a document?

Also, I assume this uses the parameter caching so the instance of Random is created once and is kept the same through the execution of the script on each doc? If not, then maybe we should do this as it will make searches with the same seed reproducible?

Comment (colings86) [reply docs/reference/query-dsl/script-score-query.asciidoc]: Actually I guess this can't do the caching since we are not using our own function. I think we should implement our own random function that caches the instance of `Random` with the provided seed?

**Review by colings86 — COMMENTED** [review]
My only comment is around random scores and whether we should have a function that caches the instance of `Random` so the same instance can be used for the whole request across all the documents. Other than that I think this is ready

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java [hunk 11]
```diff
@@ -84,6 +85,11 @@ public ScoreScript(Map<String, Object> params, SearchLookup lookup, LeafReaderCo
         return params;
     }
 
+    /** The leaf lookup for the Lucene segment this script was created for. */
+    protected final LeafSearchLookup getLeafLookup() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java]: Is this accessed from somewhere? I don't see any uses in this PR.

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java [hunk 12]
```diff
@@ -34,7 +35,7 @@
 /**
  * A script used for adjusting the score on a per document basis.
  */
-public abstract class ScoreScript {
+public abstract class ScoreScript implements ScorerAware {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java]: What's the reason for needing to add ScorerAware?

**Review by rjernst — COMMENTED** [review]
I just have 2 small last questions

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java (hunk 11 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/script/ScoreScript.java]: @rjernst  Thanks Ryan, I was copying this from a former `SearchScript`, this is removed now

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java (hunk 12 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/script/ScoreScript.java]: @rjernst Thanks Ryan, I was copying this from a former SearchScript, this is removed now

**Comment by mayya-sharipova** [comment]
@colings86  I have added random functions as you suggested, can you please continue to review when you have available time.

**Review by colings86 — APPROVED** [review]
LGTM

**Comment by MirAkram** [comment]
mj1265849@gmail.com
در تاریخ ۲۰ نوامبر ۲۰۱۸ ۷:۰۴ بعدازظهر، "Colin Goodheart-Smithe" <
notifications@github.com> نوشت:

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java [hunk 13]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java]: errant newline?

**Review by rjernst — APPROVED** [review]
LGTM
